### PR TITLE
Voucherify - Contentful

### PR DESCRIPTION
**Pull request by weronika-kurczyna**
## Purpose

We are Voucherify and we would like to have our app in your Marketplace. As we established earlier, we created PR with our application. 

With the Custom app, you can easily connect the content created in Contentful to Voucherify resources, enhancing the customer journey and your marketing teams' experience by:
- Adding links to "terms and conditions" to your campaigns so that customers know what they need to do to receive the discounts. The terms and conditions might be similar or the same for different campaigns, making it easier for the marketing team to manage it in one place.
- Adding banners and rules explanations to the Earning Rules so that the customers know how they can earn points. In case a banner or a logo changes, the change will be visible after updating the content in Contentful.
- Adding banners and promotion details so that your customers know when they are eligible for the discounts.

## Approach

The Voucherify application includes:
- (ConfigScreen) The configuration necessary for the application to work properly.
  - It is required to fill in three fields for the application to connect to Voucherify - `Application ID`, `Secret Key` and `Custom URL`. Within the Configuration Screen, URL validation has been introduced, as well as key verification.
  
image

- (FieldView) A place where resources connected to entry are visible. There are three types of Voucherify resources: Campaigns, Earning Rules and Promotion Tiers. If there are connected resources, it is possible to remove the connection between the entry and a specific resource from this level. 
image

- (DialogView) Dialog allows you to search and connect the selected resource to the entry.
It is possible to connect multiple resources to one entry, as well as to connect multiple entries to one Voucherify resource.
image

#### Voucherify
On the Voucherify side, you will need to generate API integration keys for Contentful.

- Go to Project Settings and in the General tab, scroll down to Integration Keys. 
- Click the plus button on the right.
- Fill in the Name for your key.
- Choose your role from the Role drop-down list.
- Choose Contentful from the Integration drop-down list.
- Click Create Integration API Key button.

#### Contentful - how to configure the application locally

- Go to Apps -> Create App.
- Fill in the Name for your app.
- Set URL to the `localhost:port` if you are starting the app locally from the repository or drag and drop the `build` directory from `apps/voucherify`
- In the Locations section, check the App configuration screen box.
- Check the Entry field and the JSON object.
- Click the Save button.
- Go to Apps, then select your app and install it to space.
- On the config screen fill in all the configuration fields.
 - The Custom URL should look like this: `https://cms-integrations.voucherify.io`. If your project runs on a different server region than EU1, the URL should have the prefix: `https://<region>.cms-integrations.voucherify.io`, for example: `https://us1.cms-integrations.voucherify.io`.
 - Click the `Verify Credentials` just to check if the keys and URL are valid or click the  `Install` button right away.
 
 - when the configuration is successful, create your own Content type. To enable displaying the Custom app in your entry, go to the Settings of the Content field, click Appearance and change the display option to your app to allow the entry to be linked to Voucherify.
 
image

- At this point you can create your own Entry and connect the Voucherify resources: Campaigns, Earning Rules, and Promotion Tiers to the entries.
Once integrated, Contentful automatically creates a new metadata array for linked Voucherify campaigns, earning rules, or promotion tiers upon the first connection. Each `contentfulEntity` contains the `Entry ID` and the `contentType`. If a Voucherify resource is linked to several Contentful entries, all those entries will be listed in metadata as `contentfulEntities` with respective `Entry IDs` and `contentTypes`.

Note: after updating the application configuration (if you change the project URL or credentials), you may need to refresh the page to see the current results.
 
## Breaking Changes

There are no breaking changes, we add this application for the first time.

### CODE REVIEW

**Comment by p-zielinski** [comment]
Screenshot 2023-09-11 at 11 02 04
@weronika-kurczyna
After the recent changes in Voucherify, links does not work properly. Please fix it before merge.

**Comment by weronika-kurczyna** [comment]
Hi :) One more question. Using a custom App before, you had to choose which Location to use. I know that when installing apps from the Marketplace there is no way to select this configuration, will this somehow be set automatically at the time of installation?

image

File: apps/voucherify/package.json
Comment (whitelisab) [inline apps/voucherify/package.json:11]: You could remove the `upload` and `upload-ci` scripts as they are redundant with the `deploy` scripts.

File: apps/voucherify/package.json
Comment (whitelisab) [inline apps/voucherify/package.json:14]: Can you update the app id for this `deploy:staging` script so that it will update our staging app (id: `5kadpWSYKnUZrW8ys6uvkx`)? 

```suggestion
    "deploy:staging": "contentful-app-scripts upload --ci --bundle-dir ./build --organization-id ${TEST_ORG_ID} --definition-id 5kadpWSYKnUZrW8ys6uvkx --token ${TEST_CMA_TOKEN}"
```

File: apps/voucherify/package.json
Comment (whitelisab) [inline apps/voucherify/package.json:13]: Just verifying that `6xt1AyXyw6aSnn5JDXaFin` is the app id for the Voucherify custom app that you have in your org?

File: apps/voucherify/src/components/locations/DialogView.jsx
Comment (whitelisab) [inline apps/voucherify/src/components/locations/DialogView.jsx:13]: Just a small comment here, when I was toggling between radio buttons in the dialog, I saw the React error `A component is changing an uncontrolled input to be controlled`. If you set this state as an empty string, that should fix it 😄 

```suggestion
  const [selectedResourceId, setSelectedResourceId] = useState('');
```

**Review by whitelisab — CHANGES_REQUESTED** [review]
Thank you for the detailed overview in this PR and README! I tested the app and it works really well and looks great.

To address your question about the app locations, when an app is installed from the marketplace, it is associated with a single app Id. So whatever locations are selected on that app Id is how the app will install in any other spaces. So that's why I was asking you to double check the app Id in the `deploy` script, as that will be the app Id associated with the marketplace app. 

I had some small comments in here for you to address. Please let me know if you have any questions!

File: apps/voucherify/package.json
Comment (weronika-kurczyna) [reply apps/voucherify/package.json:13]: I made a mistake, this id is `qLqU3Zs2d5awizzTcZ1mK`. :) 
This is the id of the Custom App that we created within the organization "Rspective" and space "Voucherify". Within this organization and space we built and tested our custom app. Please keep in mind that this app (its URL is: `https://cms-integrations.voucherify.io/frontend/contentful`) is slightly different from what is included in the PR, especially when it comes to the appearance and operation of the Configuration Screen.
image

File: apps/voucherify/package.json
Comment (weronika-kurczyna) [reply apps/voucherify/package.json]: Should it be the same id (5kadpWSYKnUZrW8ys6) in `deploy`? Or should I leave the id of our custom app there? (qLqU3Zs2d5awizzTcZ1mK)

File: apps/voucherify/LICENSE
Comment (whitelisab) [reply apps/voucherify/LICENSE]: @weronika-kurczyna could you update this license? Otherwise everything else looks great!

**Review by whitelisab — APPROVED** [review]
Looks great, thank you for making the changes! 

We may still update the App Id for the `deploy` script, but that will only be triggered when we push these changes to the production branch. But for now, this looks great and is ready to be merged to the main branch so we can test in our staging environment!

**Comment by weronika-kurczyna** [comment]
> Looks great, thank you for making the changes!
> 
> We may still update the App Id for the `deploy` script, but that will only be triggered when we push these changes to the production branch. But for now, this looks great and is ready to be merged to the main branch so we can test in our staging environment!

Great, I'm very happy about that, then we are waiting for the merge and further information about release.🙂

File: apps/voucherify/package.json
Comment (whitelisab) [reply apps/voucherify/package.json]: @weronika-kurczyna Sorry, one more change here. Could you update the app id in the deploy script to this `3HFf3DxTPYMh2L8214oW1s`? As mentioned in our email, we are going to use a fresh app definition for this.

Comment (gavinmatt) [reply apps/voucherify/package.json]: @weronika-kurczyna Just FYI - I made this change just to make it easier to get this PR done for release tomorrow!

**Pull request merged**
